### PR TITLE
refactor(abstract-query): improve _groupJoinData perf and readability

### PR DESCRIPTION
**Pull request by SimonSchick**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [ ] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Did you update the typescript typings accordingly (if applicable)?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md)?

### Description of change

This cleans up the probably worst function in the entire code base and manages to also make it slightly faster in the process.

Worth mentioning:
 - All code related to `getUniqueKeyAttributes` was removed as it was effectively broken (the does something that makes absolutely no sense and won't ever have an effect), I want to fix this but I don't understand what it was suppose to a achieve at all.
 - Function is about 10% faster.
 -

### CODE REVIEW

File: lib/dialects/abstract/query.js [hunk 1]
```diff
@@ -7,6 +7,46 @@ const Dot = require('dottie');
...
 const uuid = require('uuid/v4');
 
+const stringify = obj => obj instanceof Buffer ? obj.toString('hex') : obj;
+
+// Calculates the array prefix of a key (['User', 'Results'] for 'User.Results.id')
+const makeDoublePrefixMemoizer = () => {
+  const keyPrefixMemo = {};
+  return str => {
+    // We use a double memo and keyPrefixString so that different keys with the same prefix will receive the same array instead of differnet arrays with equal values
+    if (!Object.prototype.hasOwnProperty.call(keyPrefixMemo, str)) {
```
Comment (sushantdhiman) [inline lib/dialects/abstract/query.js:17]: For better readability

```Reflect.apply(Object.hasOwnProperty, keyPrefixMemo, [str]```

**Comment by sushantdhiman** [comment]
This code segment does indeed need this refactor but it is performance critical.

> Function is about 10% faster.

Any benchmark script to support this?

File: lib/dialects/abstract/query.js [hunk 2]
```diff
@@ -7,6 +7,46 @@ const Dot = require('dottie');
...
+  };
+};
+
+/**
+ * Calculate the last item in the array prefix ('Results' for 'User.Results.id')
+ *
+ * @param {string} str
+ */
+const secondSuffix = str => {
+  const lastDelim = str.lastIndexOf('.');
```
Comment (sushantdhiman) [inline lib/dialects/abstract/query.js:34]: Will this work?

```js
  if (!str.includes('.')) return '';

  const array = str.split('.');
  return array[array.length - 1];
```

Comment (SimonSchick) [reply lib/dialects/abstract/query.js:34]: I tried this approach and it was significantly slower.

File: lib/dialects/abstract/query.js (hunk 1 above)
Comment (SimonSchick) [reply lib/dialects/abstract/query.js:17]: This feels rather subjective, if anything I'd like to generate all usage of `Object.prototype.hasOwn...` to a util called `Utils.hasOwnProperty(target, str)` but this is out of scope of this PR.
The temporary array is also likely to incur another performance hit.

I'll do some benchmarking around this, I never really worked with the `Reflect` library, it feels like a clone of `Object`.

**Comment by SimonSchick** [comment]
@sushantdhiman I grabbed the the unit tests and threw them in a simple test suite `using console.time/timeEnd`, the drop was around 10% each time, if you want I can attach the benchmark here tomorrow.

**Comment by sushantdhiman** [comment]
@SimonSchick Got it, I am more concerned about unique attributes thingi. I will try to test how it was used etc

**Comment by SimonSchick** [comment]
Thanks, I was counting on you to look into that, me and @eseliger couldn't figure out what the original author was trying to achieve, the code just didn't make any sense what so ever.

File: lib/dialects/abstract/query.js (hunk 1 above)
Comment (SimonSchick) [reply lib/dialects/abstract/query.js:17]: Naive benchmark:
```js
'use strict';

const obj = {
};

function randIndex() {
  return Math.floor(Math.random() * 100);
}

for (let i = 0; i < 100;++i) {
  obj[randIndex()] = Math.random();
}

let out;
console.time('Reflect.apply');
for (let i = 0; i < 0xffffff; i++) {
  out = Reflect.apply(Object.hasOwnProperty, obj, [randIndex()]);
}
console.timeEnd('Reflect.apply');

console.time('Object.prototype.hasOwnProperty.call');
for (let i = 0; i < 0xffffff; i++) {
  out = Object.prototype.hasOwnProperty.call(obj, randIndex());
}
console.timeEnd('Object.prototype.hasOwnProperty.call');

const bound = Object.prototype.hasOwnProperty;
function hasOwnProp(obj, prop) {
  return bound.call(obj, prop);
}

console.time('hasOwnProp');
for (let i = 0; i < 0xffffff; i++) {
  out = hasOwnProp(obj, randIndex());
}
console.timeEnd('hasOwnProp');

console.log(out);
```

```
Reflect.apply: 547.062ms
Object.prototype.hasOwnProperty.call: 353.256ms
hasOwnProp: 343.112ms

Reflect.apply: 544.96ms
Object.prototype.hasOwnProperty.call: 353.499ms
hasOwnProp: 357.777ms

Reflect.apply: 547.306ms
Object.prototype.hasOwnProperty.call: 354.269ms
hasOwnProp: 350.131ms
```

Reflect is quite a bit slower than the Object variant, the wrapped (tidy) version is sometimes faster, sometimes slower but pretty much the same speed on average, I will likely file a PR to replace those later.

**Comment by SimonSchick** [comment]
@sushantdhiman do we want to move forward with this, did you find anything related to the unique checks?

**Comment by sushantdhiman** [comment]
@SimonSchick yes I want to merge this PR, will look over unique attribute case tomorrow (weekend)

**Comment by SimonSchick** [comment]
Also speed:

```
Before:
Suite 1: 4.774s
Suite 2: 4.326s
Suite 3: 4.814s
Suite 4: 4.142s
Suite 5: 3.155s
Suite 6: 3.756s
Suite 7: 2.079s

After:
Suite 1: 4.468s
Suite 2: 4.076s
Suite 3: 4.370s
Suite 4: 3.943s
Suite 5: 2.921s
Suite 6: 3.377s
Suite 7: 1.640s
```

https://gist.github.com/SimonSchick/aea921ec33879833b19b6cb198d2f375
All the suites are based on the existing test cases

**Comment by eseliger** [comment]
That's awesome!

**Comment by sushantdhiman** [comment]
@SimonSchick This method is used to populate results when included model got no PK. But it is currently broken. It relies on `model.attributes` incorrectly, it should be using `model.rawAttributes`

```diff
const getUniqueKeyAttributes = model => {
  let uniqueKeyAttributes = _.chain(model.uniqueKeys);
  uniqueKeyAttributes = uniqueKeyAttributes
    .result(`${uniqueKeyAttributes.findKey()}.fields`)
-   .map(field => _.findKey(model.attributes, chr => chr.field === field))
+   .map(field => _.findKey(model.rawAttributes, chr => chr.field === field))
    .value();

  return uniqueKeyAttributes;
};
```

I am including a test case below with correct results. It is broken on

1) `master`, because `model.attributes` is used to find unique fields
2) `this branch`, because it completely removed handling of unique non-PK based models 

```js
(async () => {
  const Foo = sequelize.define("Foo", {
    name: {
      type: Sequelize.STRING,
      unique: true,
    },
  });

  // Foo.removeAttribute("id");

  const Bar = sequelize.define("Bar", {
    name: {
      type: Sequelize.STRING,
      unique: true,
    },
  });

  Bar.removeAttribute("id");

  Foo.hasMany(Bar, { foreignKey: "fooName", sourceKey: "name" });
  Bar.belongsTo(Foo, { foreignKey: "fooName", sourceKey: "name" });

  await sequelize.sync({ force: true });
  await Foo.bulkCreate([{ name: "foo-a" }, { name: "foo-b" }]);
  await Bar.bulkCreate([{ name: "bar-a", fooName: "foo-a" }]);
  await Bar.bulkCreate([{ name: "bar-b", fooName: "foo-a" }]);

  console.log(
    JSON.stringify(await Foo.findAll({ include: { all: true } }), null, 2)
  );
})();
```

```sql
[
  {
    "id": 1,
    "name": "foo-a",
    "createdAt": "2020-05-09T06:41:22.289Z",
    "updatedAt": "2020-05-09T06:41:22.289Z",
    "Bars": [
      {
        "name": "bar-a",
        "createdAt": "2020-05-09T06:41:22.294Z",
        "updatedAt": "2020-05-09T06:41:22.294Z",
        "fooName": "foo-a"
      },
      {
        "name": "bar-b",
        "createdAt": "2020-05-09T06:41:22.295Z",
        "updatedAt": "2020-05-09T06:41:22.295Z",
        "fooName": "foo-a"
      }
    ]
  },
  {
    "id": 2,
    "name": "foo-b",
    "createdAt": "2020-05-09T06:41:22.289Z",
    "updatedAt": "2020-05-09T06:41:22.289Z",
    "Bars": [
      {
        "name": null,
        "createdAt": null,
        "updatedAt": null,
        "fooName": null
      }
    ]
  }
]
```

When issue pointed in this comment is addressed, https://github.com/sequelize/sequelize/issues/5193 will be fixed as well :tada:

**Comment by SimonSchick** [comment]
This still doesn't make sense with `.result(`${uniqueKeyAttributes.findKey()}.fields`}` as this only returns the first item of the `uniqueKeyAttributes`?

**Comment by sushantdhiman** [comment]
> This still doesn't make sense with .result(${uniqueKeyAttributes.findKey()}.fields} as this only returns the first item of the uniqueKeyAttributes?

AFAIU, in case of multiple unique keys it work with the first available unique key. It make sense as records will be unique for any of `uniqueKeys` selection, we can use any of those keys to find unique record.

**Comment by SimonSchick** [comment]
Alright, I will re-add this logic, thanks for looking into it.

**Comment by SimonSchick** [comment]
Will this not fail for compound unique keys though? See
`_.findKey(model.attributes, chr => chr.field === field)` compares fields by reference which might be different arrays?

**Comment by sushantdhiman** [comment]
> _.findKey(model.attributes, chr => chr.field === field)

I think this logic is trying to convert field names to their attribute name, `uniqueKey.fields` are usually column names, which may not match with their respective attribute name in model

**Comment by eseliger** [comment]
@SimonSchick Addressed Sushant's feedback. So this closes #5193.

**Pull request closed without merging**